effect passing chain

secminhr opened the pull request:
Use effect passing chain to remove the dependency from player to effect.

secminhr committed:
Make game can start running (in a better way).

secminhr committed:
Separate effect Summon and Player

secminhr committed:
Effect Handling Chain facilities.

secminhr committed:
Separate effects and player.

secminhr committed:
Adapt all cards to new effect system.

secminhr committed:
Change EffectFunctions description.

secminhr committed:
Adapt all Characters to new effect system.

secminhr committed:
Merging

secminhr commented:
The passive skills aren't considered during the design of this.
They will be included in future PRs.

aoaaceai commented on src/main/java/eod/Game.java:
```diff
@@ -125,6 +131,15 @@ public void start() {
 
+    private Round changeRound(Round current) {
+        int currentNumber = current.getNumber();
+        if(current.getPlayer().equals(playerOrder[0])) {
```
use player.isPlayerA() since .equals() takes time

secminhr replied:
fixed

aoaaceai commented on src/main/java/eod/card/concrete/command/EquivalentExchange.java:
```diff
@@ -15,9 +17,14 @@ public EquivalentExchange() {
+        Leader leader = player.getLeader();
+        Effect[] effects = new Effect[] {
+            //An attack to itself, then the handler should be its owner
+            RequestDirectAttack(2, Effect.HandlerType.Owner).from(leader).to(leader),
```
self-attacking can cause FightBack to malfunction

aoaaceai replied:
and also the FURIOUS effect.

secminhr replied:
fixed

aoaaceai commented on src/main/java/eod/card/concrete/normal/GrandFeast.java:
```diff
@@ -55,6 +55,7 @@ public void onEventOccurred(GameObject sender, Event event) {
         if (event instanceof AttackEvent) {
             AttackEvent e = (AttackEvent) event;
             if(e.getSender().isPlayerA() == player.isPlayerA()) {
+                //TODO: don't think this will really work, need tests and fixes if necessary
```
this works, but unfortunately Spacezipper changed the description of GrandFeast. The next PR will fix that.

secminhr replied:
removed

aoaaceai commented on src/main/java/eod/card/concrete/normal/SingleContract.java:
```diff
@@ -16,8 +17,12 @@ public SingleContract() {
-    public void applyEffect() {
+    public void applyEffect(EffectExecutor executor) {
         WarObject[]  characters = WarObject(player.getBoard()).which(OwnedBy(player.rival())).which(Being(Character.class)).which(WithoutStatus(Status.SNEAK)).get();
+        //TODO: adapt the attack to effect
```
Maybe we should add a DamageEffect

secminhr replied:
The biggest problem is that the effect is given by rules, which means no attacker.
It will be fixed in future PRs.

secminhr replied:
Fixed

aoaaceai commented on src/main/java/eod/effect/DirectAttack.java:
```diff
@@ -6,17 +6,26 @@
 
-    public DirectAttack(Player player, int hp) {
-        super(player, hp);
+    //Use HandlerType in constructor because a character might attack itself,
```
it wont attack itself anymore.

aoaaceai commented on src/main/java/eod/warObject/character/concrete/red/GuerrillaShooter.java:
```diff
@@ -36,16 +37,16 @@ public String getName() {
-        ArrayList<Damageable> affected = RequestRegionalAttack(player, attack).from(this).to(getAttackRange(), 1).getAffected();
+    public void attack(EffectExecutor executor) {
+        super.attack(executor);
+        ArrayList<Damageable> affected = RequestRegionalAttack(attack).from(this).to(player, getAttackRange(), 1).getAffected();
```
affected should contain nothing cuz RegionalAttack.to() doesn't *really* attack the targets. (executor missing)

secminhr replied:
fixed

aoaaceai commented on src/main/java/eod/warObject/character/concrete/red/LeadersGuard.java:
```diff
@@ -32,10 +34,10 @@ public String getName() {
+    public void attack(EffectExecutor executor) {
+        super.attack(executor);
+        SpecialRegionalAttack SRA = (SpecialRegionalAttack) RequestRegionalAttack(attack).from(this);
+        SRA.to(player, getAttackRange(), 1);
```
executor missing

secminhr committed:
first comment
do while exit condition fix

aoaaceai commented on src/main/java/eod/warObject/character/concrete/red/ToughGuy.java:
```diff
@@ -31,9 +31,9 @@ public String getName() {
-        RequestRegionalAttack(player, attack).from(this).to(getAttackRange());
+    public void attack(EffectExecutor executor) {
+        super.attack(executor);
+        RequestRegionalAttack(attack).from(this).to(getAttackRange());
```
executor missing

aoaaceai commented on src/main/java/eod/warObject/character/concrete/red/MafiaAssassin.java:
```diff
@@ -36,10 +35,10 @@ public String getName() {
+    public void attack(EffectExecutor executor) {
+        super.attack(executor);
+        SpecialRegionalAttack a = new SpecialRegionalAttack(attack);
+        a.from(this).to(player, getAttackRange(), 1);
```
executor missing

secminhr committed:
comment 3
add explanation on SingleContract.applyEffect

secminhr committed:
comment 2
comment 4
comment 5
Add Damage, which no attacker is specified.
Fix DirectAttack.
Add Player.damage
Change original Player.damage to Player.attack

secminhr committed:
Last comment.
